File: app-java/src/test/java/tech/thdev/java_udemy_sample/class_02/_01_Abstract.java
package tech.thdev.java_udemy_sample.class_02;

/**
 * Created by tae-hwan on 09/10/2016.
 */

/**
 * Java에서는 모든 함수가 재정의가 가능하다
 */
public abstract class _01_Abstract {

    public String attributeA;
    public String attributeB = "AttributeB";

    public abstract void methodA();

    /**
     * final을 사용할 경우 함수 재정의가 불가능하다
     */
    public final String methodB() {
        return attributeA;
    }
}

/**
 * final을 사용하였기에 다음의 클래스는 재정의가 불가능합니다
 */
final class AbstractSample extends _01_Abstract {

    @Override
    public void methodA() {
        attributeA = "ABC";

        System.out.println("methodA");
    }

    /**
     * final로 감싸두어서 Override를 할 수 없다
     */
//    @Override
//    public String methodB() {
//        return "";
//    }
}

File: app-kotlin/src/test/java/tech/thdev/kotlin_udemy_sample/class_02/_01_abstract.kt
package tech.thdev.kotlin_udemy_sample.class_02

/**
 * Created by tae-hwan on 09/10/2016.
 */

abstract class _01_Abstract {

    abstract var attributeA: String

    // 상속 후 재 정의를 하지 못함
    val attributeB: String = "AttributeB"

    abstract fun methodA()

    // 상속 후 재 정의를 하지 못함
    /*open */fun methodB(): String {
        return attributeA
    }
}

class AbstractSample : _01_Abstract() {

    override var attributeA: String = "AttributeA"

    override fun methodA() {
        attributeA = "Sample"

        println("---------")
        println("methodA")
        println("---------")
    }

    /**
     * Method B에대해서 method 재정의를 허용하고 싶다면 `open` 키워드를 추가해야 합니다
     */
//    override fun methodB(): String {
//        return super.methodB()
//    }
}